File: src/Butil/Bit.Butil/Publics/AbortController.cs
using System;
using System.Collections.Concurrent;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps <see href="https://developer.mozilla.org/en-US/docs/Web/API/AbortController">AbortController</see>
/// and <see href="https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal">AbortSignal</see>:
/// a cancellation token the browser understands, which you create once and hand to as many
/// operations as you like.
/// </summary>
/// <remarks>
/// The point of a standalone controller is sharing. A <see cref="ButilAbortSignal"/> can be passed
/// to several APIs at once, and one <see cref="AbortControllerHandle.Abort"/> cancels all of them -
/// which the per-API abort handles (<see cref="AbortableFetch"/> and friends) cannot do, since each
/// only reaches its own call.
/// <br/>
/// <see cref="Timeout"/> and <see cref="Any"/> produce signals nobody can abort by hand: the first
/// fires itself after a delay, the second when any of its sources fires. Composing them - a request
/// deadline that a Cancel button can also trip - is what they are for.
/// </remarks>
[ButilService(typeof(AbortController))]
public class AbortController(IJSRuntime js) : IAsyncDisposable
{
    internal const string AbortMethodName = nameof(InvokeAbort);

    // Keyed by listener id, but carrying the signal it belongs to: releasing a signal has to drop
    // its listeners, and the listener id is all the abort dispatch itself has to go on.
    private readonly ConcurrentDictionary<Guid, (Guid SignalId, Action<string> OnAbort)> _handlers = new();

    // Per-instance callback reference: signals are isolated per circuit / WASM app and released on
    // disposal - no static state, no cross-circuit leak.
    private DotNetObjectReference<AbortController>? _dotNetRef;
    private DotNetObjectReference<AbortController> DotNetRef => DotNetObjectReferenceHelper.GetOrCreate(ref _dotNetRef, this);

    /// <summary>True when the runtime exposes <c>AbortController</c>.</summary>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (<c>false</c>) rather than
    /// throwing, so the result can't be distinguished from a genuine value. If you branch on it,
    /// defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.abortController.isSupported");

    /// <summary>True when the runtime exposes <c>AbortSignal.timeout</c>.</summary>
    public ValueTask<bool> IsTimeoutSupported() => js.Invoke<bool>("BitButil.abortController.isTimeoutSupported");

    /// <summary>
    /// True when the runtime exposes <c>AbortSignal.any</c>. <see cref="Any"/> still works when this
    /// is false - it falls back to a hand-wired controller with the same behaviour.
    /// </summary>
    public ValueTask<bool> IsAnySupported() => js.Invoke<bool>("BitButil.abortController.isAnySupported");

    /// <summary>
    /// Invoked from JS when a signal aborts. Public + <see cref="JSInvokableAttribute"/> so it can be
    /// dispatched through the per-instance <see cref="DotNetObjectReference{T}"/>.
    /// </summary>
    [JSInvokable(AbortMethodName)]
    public void InvokeAbort(Guid listenerId, string reason)
    {
        if (_handlers.TryGetValue(listenerId, out var handler)) handler.OnAbort(reason);
    }

    /// <summary>
    /// Creates a controller and its signal.
    /// </summary>
    /// <returns>A handle, or null when the runtime has no <c>AbortController</c>.</returns>
    /// <remarks>
    /// Dispose the handle when you're done to release the signal's JS entry. Disposing does not
    /// abort it - releasing a handle is not the same as cancelling what it guards.
    /// </remarks>
    public async ValueTask<AbortControllerHandle?> Create()
    {
        var id = Guid.NewGuid();
        var created = await js.Invoke<bool>("BitButil.abortController.create", id);

        return created ? new AbortControllerHandle(js, this, id) : null;
    }

    /// <summary>
    /// Creates a signal that aborts itself after <paramref name="delay"/> with a
    /// <c>TimeoutError</c> reason. Nothing can abort it early.
    /// </summary>
    /// <param name="delay">How long to wait. Rounded to whole milliseconds. Must not be negative.</param>
    /// <returns>A signal, or null when the runtime has no <c>AbortSignal.timeout</c>.</returns>
    /// <exception cref="ArgumentOutOfRangeException"><paramref name="delay"/> is negative.</exception>
    /// <remarks>
    /// The timer starts immediately, not when the signal is first used. Combine it with a
    /// controller's own signal through <see cref="Any"/> to get a deadline that a user can also
    /// cancel by hand.
    /// </remarks>
    public async ValueTask<ButilAbortSignal?> Timeout(TimeSpan delay)
    {
        // AbortSignal.timeout takes an [EnforceRange] unsigned long long, so a negative delay - the
        // shape Timeout.InfiniteTimeSpan arrives in - is a TypeError in the browser rather than a
        // deadline that never fires. Refuse it here, where the caller can see which argument it was.
        ArgumentOutOfRangeException.ThrowIfLessThan(delay, TimeSpan.Zero);

        var id = Guid.NewGuid();
        var created = await js.Invoke<bool>("BitButil.abortController.timeout", id, (long)delay.TotalMilliseconds);

        return created ? new ButilAbortSignal(js, this, id) : null;
    }

    /// <summary>
    /// Creates a signal that aborts as soon as any of <paramref name="signals"/> does, carrying that
    /// source's reason.
    /// </summary>
    /// <returns>
    /// A signal, or null when a source is null or has already been released, or the runtime has
    /// neither <c>AbortSignal.any</c> nor <c>AbortController</c>.
    /// </returns>
    /// <remarks>
    /// The composite does not keep its sources alive: release one and the composite is left watching
    /// a signal that can no longer fire, so dispose the composite first. Where <c>AbortSignal.any</c>
    /// is missing (Safari before 17.4) this wires a controller to the sources by hand, which behaves
    /// identically.
    /// </remarks>
    public async ValueTask<ButilAbortSignal?> Any(params ButilAbortSignal[] signals)
    {
        ArgumentNullException.ThrowIfNull(signals);

        // A null source is the same kind of answer as a released one: there is no signal to compose,
        // and a composite missing one of its sources would silently be weaker than it was asked for.
        if (Array.Exists(signals, s => s is null)) return null;

        var id = Guid.NewGuid();
        var created = await js.Invoke<bool>("BitButil.abortController.any", id, Array.ConvertAll(signals, s => s.Id));

        return created ? new ButilAbortSignal(js, this, id) : null;
    }

    internal async ValueTask<ButilSubscription?> AddAbortListener(Guid signalId, Action<string> onAbort)
    {
        ArgumentNullException.ThrowIfNull(onAbort);

        var listenerId = Guid.NewGuid();
        _handlers[listenerId] = (signalId, onAbort);

        var added = await js.Invoke<bool>("BitButil.abortController.addListener", DotNetRef, signalId, listenerId);

        if (added is false)
        {
            _handlers.TryRemove(listenerId, out _);
            return null;
        }

        return new ButilSubscription(listenerId, async () =>
        {
            _handlers.TryRemove(listenerId, out _);
            await js.InvokeVoid("BitButil.abortController.removeListener", signalId, listenerId);
        });
    }

    // A released signal can never fire again, so every listener still registered for it is dead
    // weight: the JS release drops its half, and this drops the .NET half it cannot reach - a
    // subscription the caller never disposed would otherwise be held for the life of the service.
    internal void ForgetSignalListeners(Guid signalId)
    {
        foreach (var pair in _handlers)
        {
            if (pair.Value.SignalId == signalId) _handlers.TryRemove(pair.Key, out _);
        }
    }

    /// <summary>
    /// On scope/circuit teardown, releases every signal whose handle was never disposed, along with
    /// the interop reference their callbacks were dispatched through.
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        try
        {
            _handlers.Clear();
            await js.InvokeVoid("BitButil.abortController.releaseAll");
        }
        catch (Exception ex) when (ex.IsIgnorableDisposalException()) { } // teardown: circuit gone, cancelled, or already disposed
        finally
        {
            _dotNetRef?.Dispose();
            _dotNetRef = null;
        }
        GC.SuppressFinalize(this);
    }
}

File: src/Butil/Bit.Butil/Scripts/abortController.ts
var BitButil = (window as any).BitButil = (window as any).BitButil || {};

(function (butil: any) {
    interface Entry {
        signal: AbortSignal;
        controller?: AbortController;   // absent for timeout()/any()-derived signals, which nobody can abort directly
        listeners: { [listenerId: string]: (e: any) => void };
        // Only for a hand-wired composite (the pre-Safari-17.4 any() fallback): the source signals
        // it is watching, with the listener each one is holding on its behalf.
        sources?: { source: AbortSignal; handler: () => void }[];
    }

    // Every signal Butil hands to .NET lives in the abortSignals registry under the id .NET knows
    // it by. Other modules (fetch, scheduler) reach in there rather than through this module, so a
    // request that composes a shared signal does not download the whole AbortController surface.

    // AbortSignal.reason is any value at all - a DOMException for abort()/timeout(), whatever was
    // passed otherwise. .NET only ever sees a string, so flatten it here in one place.
    function reasonOf(signal: AbortSignal): string {
        const reason: any = (signal as any).reason;
        if (reason === undefined || reason === null) return '';
        if (reason instanceof DOMException) return reason.message || reason.name;
        if (typeof reason === 'string') return reason;
        try { return JSON.stringify(reason); } catch { return String(reason); }
    }

    function track(id: string, signal: AbortSignal, controller?: AbortController) {
        butil.abortSignals.put(id, { signal, controller, listeners: {} });
    }

    // A source signal can outlive the composite watching it by a long way - a timeout signal held
    // for the life of the page, say - and until its listener is removed it keeps that composite
    // alive with it. Detach them as a set: the first abort makes the rest pointless anyway.
    function detachSources(entry: Entry) {
        if (!entry.sources) return;
        for (const pair of entry.sources) pair.source.removeEventListener('abort', pair.handler);
        entry.sources = undefined;
    }

    butil.abortController = {
        isSupported() { return typeof (window as any).AbortController === 'function'; },

        // True only when AbortSignal.any exists - Any() falls back to a controller wired by hand,
        // which behaves the same but cannot be told apart from a real composite signal.
        isAnySupported() { return typeof (window as any).AbortSignal?.any === 'function'; },
        isTimeoutSupported() { return typeof (window as any).AbortSignal?.timeout === 'function'; },

        create(id: string) {
            if (typeof (window as any).AbortController !== 'function') return false;
            const controller = new AbortController();
            track(id, controller.signal, controller);
            return true;
        },

        timeout(id: string, milliseconds: number) {
            const AS: any = (window as any).AbortSignal;
            if (typeof AS?.timeout !== 'function') return false;
            track(id, AS.timeout(milliseconds));
            return true;
        },

        // Composite: aborts as soon as any source does, carrying that source's reason.
        any(id: string, sourceIds: string[]) {
            const sources = (sourceIds ?? []).map(sourceId => butil.abortSignals.signalOf(sourceId)).filter(Boolean) as AbortSignal[];
            if (sources.length !== (sourceIds ?? []).length) return false; // an unknown id would silently weaken the composite

            const AS: any = (window as any).AbortSignal;
            if (typeof AS?.any === 'function') {
                track(id, AS.any(sources));
                return true;
            }

            if (typeof (window as any).AbortController !== 'function') return false;

            // Pre-Safari-17.4 fallback. A hand-wired controller is observably the same thing: it
            // forwards the first source's reason and stops listening once it has fired.
            const controller = new AbortController();
            track(id, controller.signal, controller);

            const already = sources.find(source => source.aborted);
            if (already) {
                controller.abort((already as any).reason);
                return true;
            }

            const entry = butil.abortSignals.entryOf(id);
            entry.sources = sources.map(source => ({
                source,
                handler: () => {
                    detachSources(entry);
                    controller.abort((source as any).reason);
                }
            }));
            for (const pair of entry.sources) pair.source.addEventListener('abort', pair.handler, { once: true });
            return true;
        },

        abort(id: string, reason: string | null) {
            const entry = butil.abortSignals.entryOf(id);
            if (!entry?.controller) return false;
            // abort() with no argument produces the standard "AbortError" DOMException, which is
            // what callers who pass no reason expect to see; abort('') would replace it with ''.
            try { reason ? entry.controller.abort(reason) : entry.controller.abort(); } catch { /* already aborted */ }
            return true;
        },

        aborted(id: string) { return butil.abortSignals.signalOf(id)?.aborted ?? false; },

        reason(id: string) {
            const entry = butil.abortSignals.entryOf(id);
            return entry ? reasonOf(entry.signal) : '';
        },

        addListener(dotNetRef: any, id: string, listenerId: string) {
            const entry = butil.abortSignals.entryOf(id);
            if (!entry) return false;

            const handler = () => butil.utils.dispatch(dotNetRef, 'InvokeAbort', listenerId, reasonOf(entry.signal));

            // An already-aborted signal never fires 'abort' again, so a late listener would never be
            // called at all. Dispatching once matches what a caller means by "tell me when this aborts".
            if (entry.signal.aborted) {
                handler();
                return true;
            }

            entry.listeners[listenerId] = handler;
            entry.signal.addEventListener('abort', handler, { once: true });
            return true;
        },

        removeListener(id: string, listenerId: string) {
            const entry = butil.abortSignals.entryOf(id);
            const handler = entry?.listeners[listenerId];
            if (!handler) return;
            delete entry.listeners[listenerId];
            entry.signal.removeEventListener('abort', handler);
        },

        // Drops the registry entry. The signal itself is not aborted - releasing a handle is not the
        // same as cancelling what it guards, and a signal already handed to fetch keeps working.
        release(id: string) {
            const entry = butil.abortSignals.entryOf(id);
            if (!entry) return;
            for (const listenerId of Object.keys(entry.listeners)) {
                entry.signal.removeEventListener('abort', entry.listeners[listenerId]);
            }
            detachSources(entry);
            butil.abortSignals.remove(id);
        },

        releaseAll() {
            for (const id of butil.abortSignals.ids()) butil.abortController.release(id);
        }
    };
}(BitButil));
